File: domain/src/test/java/hk/qingke/learn/usermanager/entity/UserEntityTest.java
package hk.qingke.learn.usermanager.entity;

import hk.qingke.learn.usermanager.entity.exception.PasswordIllegalException;
import hk.qingke.learn.usermanager.entity.exception.UsernameIsBlankException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

class UserEntityTest {
    @Test
    void create_user_entity_success() {
        Assertions.assertDoesNotThrow(() -> {
            new UserEntity("Tom", "1@Aa1234567", "test@test.hk");
        });
    }

    @Test
    void create_user_entity_fail_given_blank_username() {
        UserEntity userEntity = new UserEntity();
        Assertions.assertThrows(UsernameIsBlankException.class, () -> userEntity.setUsername(null));
    }

    @Test
    void create_user_entity_fail_given_valid_password() {
        List<String> illegalPasswords = Arrays.asList(
                "12345678",
                "abcdefgh",
                "ABCDEFGH",
                "@@@@@@@@",
                "1@ab",
                "1@ab cde"
        );

        for (String illegalPassword : illegalPasswords) {
            UserEntity userEntity = new UserEntity();
            Assertions.assertThrows(PasswordIllegalException.class, () -> userEntity.setPassword(illegalPassword));
        }
    }
}

File: domain/src/test/kotlin/hk/qingke/kotlin/usermanager/domain/UserEntityTest.kt
package hk.qingke.kotlin.usermanager.domain

import hk.qingke.kotlin.usermanager.domain.exception.EmailIllegalException
import hk.qingke.kotlin.usermanager.domain.exception.PasswordIllegalException
import hk.qingke.kotlin.usermanager.domain.exception.UsernameIllegalException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class UserEntityTest {
    @Test
    fun `user entity should be created without exception`() {
        Assertions.assertDoesNotThrow {
            UserEntity("tom-test-123_test", "12345678@Ab", "test@test.hk")
        }
    }

    // region username exception unit test

    @Test
    fun `should throw username illegal exception given username is less than min length`() {
        Assertions.assertThrows(UsernameIllegalException::class.java) {
            UserEntity("tom", "12345678@Ab", "test@test.hk")
        }
    }

    @Test
    fun `should throw username illegal exception given username is greater than max length`() {
        Assertions.assertThrows(UsernameIllegalException::class.java) {
            UserEntity("tom-test-too-too-long", "12345678@Ab", "test@test.hk")
        }
    }

    @Test
    fun `should throw username illegal exception given username start with number`() {
        Assertions.assertThrows(UsernameIllegalException::class.java) {
            UserEntity("1tom123", "12345678@Ab", "test@test.hk")
        }
    }

    @Test
    fun `should throw username illegal exception given username contain illegal char`() {
        Assertions.assertThrows(UsernameIllegalException::class.java) {
            UserEntity("tom-test@", "12345678@Ab", "test@test.hk")
        }
    }
    // endregion

    // region password exception unit test
    @Test
    fun `should throw password illegal exception given password just contain number`() {
        Assertions.assertThrows(PasswordIllegalException::class.java) {
            UserEntity("tom-test", "12345678", "test@test.hk")
        }
    }

    @Test
    fun `should throw password illegal exception given password is less than min length`() {
        Assertions.assertThrows(PasswordIllegalException::class.java) {
            UserEntity("tom-test", "1234@Ab", "test@test.hk")
        }
    }

    @Test
    fun `should throw password illegal exception given password is greater than max length`() {
        Assertions.assertThrows(PasswordIllegalException::class.java) {
            UserEntity("tom-test", "12345678901234567890@Ab", "test@test.hk")
        }
    }
    // endregion

    // region email exception unit test
    @Test
    fun `should throw password illegal exception given email not contain @`() {
        Assertions.assertThrows(EmailIllegalException::class.java) {
            UserEntity("tom-test-123_test", "12345678@Ab", "test.test.hk")
        }
    }

    @Test
    fun `should throw password illegal exception given email contain illegal char`() {
        Assertions.assertThrows(EmailIllegalException::class.java) {
            UserEntity("tom-test-123_test", "12345678@Ab", "test@test@test.hk")
        }
    }
    // endregion
}